### PR TITLE
WIP to fix #24100, #24067 -- Made pre and post_migrate signals dispatch the migration plan.

### PR DESCRIPTION
**Pull request by charettes**
I'd like to try to write a fix for both tickets to see if `plan` is sufficient to fix both issues.

### CODE REVIEW

File: docs/ref/signals.txt [hunk 1]
```diff
@@ -405,6 +405,13 @@ Arguments sent with this signal:
 ``using``
     The alias of database on which a command will operate.
 
+``plan``
+    .. versionadded:: 1.8
+
+        The migration plan; a list of two elements tuples containing
```
Comment (timgraham) [inline docs/ref/signals.txt]: deindent one level (so it's at the same level as `.. versionadded:: 1.8`) [and below]

File: docs/ref/signals.txt [hunk 2]
```diff
@@ -405,6 +405,13 @@ Arguments sent with this signal:
...
+        ``Migration`` instances coupled with a boolean telling whether or not
+        it will be applied backward.
```
Comment (timgraham) [inline docs/ref/signals.txt]: I think "backwards" is correct grammar.

**Comment by charettes** [comment]
Fixed your comments Tim, thanks!

While trying to write a fix for [#24075](https://code.djangoproject.com/ticket/24075) I realized that the `plan` wouldn't be enough to figure out whether or not the content type or permission table are created. We need the final `state` of the migration to figure that out correctly. Would it make sense to also dispatch this?

**Comment by timgraham** [comment]
If it seems needed, I'd say let's see what the code looks like.

**Comment by MarkusH** [comment]
Dispatching the state seems to be a requirement for #24075 in my eyes. You could construct the state from the plan, but that shouldn't be the solution.

I think the `pre_migrate` signal should receive the project state right before the migration process starts (i.e., blank database --> no apps and models), while the `post_migrate` signal should get the project state after the last migration is applied.

But before exposing any of the currently internal migrations API via a public interface I'd prefer to make sure the state rewrite by @claudep really works. I'm confident it does, but there are some edge cases that are hard to fix. Until the rewrite proves stable, which is the 1.8 release (or at least rc), I'm not convinced pushing the state to a public API is a good idea.

With respect to #24075: this is bug which could be backported to 1.8 once we focus on the 1.9 development according to our release process. I don't think 1.7 will be a problem. Many people are probably going to upgrade to 1.8 just for the sake of migration speedups.

**Comment by charettes** [comment]
@MarkusH started working on this.

**Pull request closed without merging**